### PR TITLE
pgp-packet-dumper based on librepgp

### LINKED ISSUE #442: Create a pgp-packet-dumper based on librepgp
## Description

Originally discussed in #441.

[pgpdump](http://www.mew.org/~kazu/proj/pgpdump/en/) is an OpenPGP packet dumper.

GnuPG has a packet dumper too, output can be seen here:
https://davesteele.github.io/gpg/2014/09/20/anatomy-of-a-gpg-key/

## Expected Behavior

We should have a OpenPGP packet dumper that does exactly those things, just using the newly extracted `librepgp`.

Potentially more, e.g., listing packets out in different formats such as JSON).

This will also be a great demonstration of the completeness of `librepgp`.

### PR DESCRIPTION
**Pull request by flowher**
This patch implements ``redumper`` - pgp-packet-dumper based on ``librepgp`` which is a stand alone application. The application internally uses  ``repgp_list_packets`` API. Some improvements has been added in ``print-packet.c`` in order to "prettify" an output and make packet content printing configurable.

```
Usage of redumper:
            -i input_file [mandatory]: specifies name of the file with PGP packets
            -d : indicates wether to dump whole packet content
            -h : prints help and exists
```

Example of output:
```
> ./redumper -i pubkey.pgp   

* PACKET: Secret Key (tag 5) (958 bytes) offset=0x0 format=new
       S2K Usage: 254
       S2K Specifier: 3
       Symmetric algorithm: CAST5 (0x3)
       Hash algorithm: SHA256 (0x8)
       Salt: [data size: 8 bytes]
       Octet count: 65536
       IV: [data size: 8 bytes]

* PACKET: User ID (tag 13) (54 bytes) offset=0x3c1 format=new
       userid: RSA (Encrypt or Sign) 2048-bit key <flowher@localhost>

* PACKET: Signature (tag 2) (306 bytes) offset=0x3f9 format=new
       Version: 4
       Signature Type: Positive certification of a User ID and Public Key packet (0x13)
       Public Key Algorithm: RSA (Encrypt or Sign) (0x1)
       Hash Algorithm: SHA256 (0x8)
       Hashed data len: 0
       * Signature Creation Time (type 0x02)
               Signature Creation Time: time=1504459919 (Sun Sep  3 18:31:59 2017)
       * Key Flags (type 0x1b)
               Key Flags: [data size: 1 bytes]
               .. May be used to sign data
               .. May be used to certify other keys
       * Preferred Symmetric Algorithms (type 0x0b)
               Preferred Symmetric Algorithms: [data size: 4 bytes]
               .. AES (256-bit key)
               .. AES (192-bit key)
               .. AES (128-bit key)
               .. TripleDES
       * Preferred Hash Algorithms (type 0x15)
               Preferred Hash Algorithms: [data size: 5 bytes]
               .. SHA256
               .. SHA384
               .. SHA512
               .. SHA224
               .. SHA1
       * Preferred Compression Algorithms (type 0x16)
               Preferred Compression Algorithms: [data size: 4 bytes]
               .. ZLIB(RFC1950)
               .. Bzip2(BZ2)
               .. ZIP(RFC1951)
               .. Uncompressed
       * Issuer key ID (type 0x10)
               Issuer Key Id: [data size: 8 bytes]
              hash2: [data size: 2 bytes]
       RSA sign = 14BEC45C04E8F0146B2BC67D6813487CA7EA400CFF638839D8B6C1EFCDE98B1A8E16339EDAA9612E046DECAD13A53496D65DD14B1CC1A13F9161ECD3407233CCE1C162666ABFE3065A18B462CBD2C241EBAA0EB6E48D85752C3B1DEE0F8776D994A1607FF2B5758C4534C2039D006DF06A2B05A9FD0506B5C3510113A4A404033C01730918973C76BA6EDF335DA7CF48F06AE41FCAAC03FB2C1EF5645B478DDFBBB52E244A0BB54924CF7DEAA8470D7AA8BDAEDA655D6343264B7D2215CC5CFE8845563333BCAD7F81C5A391884A7047BB07086C2B6AC5C0B02679F01F9B0625E5C7DAD409F00D9BE55E1206BAB9B0CE2134C4AAD4384395CD361A6A259420E6 (2045 bits)

```

Open for review by anyone
#309
Closes #442

### CODE REVIEW

**Comment by ronaldtse** [comment]
@flowher this looks really good 👍  So glad this is now extracted. Can you help add the usage of this in the README?

**Comment by flowher** [comment]
@ronaldtse README - definitely good idea. I'll add it today

**Comment by dewyatt** [comment]
> Why not just redumper , without the -i parameter? Other then this looks good. Also would be good to display contents of encrypted packets later.

Same thought I had. If it's a mandatory arg, it may as well be a positional arg. Also if this is anything more than a demo, it should default to stdin if nothing is provided.

**Comment by flowher** [comment]
okis dokis. Let me change it

File: src/librepgp/packet-print.c [hunk 1]
```diff
@@ -1023,143 +1005,169 @@ pgp_sprint_pubkey(const pgp_key_t *key, char *out, size_t outsize)
...
+    *indent -= 4;
+}
+
+inline static void
+print_packet_length_type(int indent, const pgp_ptag_t *ptag)
+{
+    /* Applicable only if old format */
+    if (ptag->new_format == 0) {
+        print_indent(indent);
+        printf("Length type: %d", ptag->length_type);
```
Comment (dewyatt) [inline src/librepgp/packet-print.c]: Add newline?

File: src/librepgp/packet-print.c [hunk 2]
```diff
@@ -1023,143 +1005,169 @@ pgp_sprint_pubkey(const pgp_key_t *key, char *out, size_t outsize)
...
-*/
 static void
-print_seckey_verbose(const pgp_content_enum type, const pgp_seckey_t *seckey)
+print_seckey_verbose(pgp_printstate_t *     printstate,
+                     const pgp_content_enum type,
+                     const pgp_seckey_t *   seckey)
 {
-    printf("------- SECRET KEY ------\n");
-    print_tagname(0, (type == PGP_PTAG_CT_SECRET_KEY) ? "SECRET_KEY" : "SECRET_SUBKEY");
     /* pgp_print_pubkey(key); */
```
Comment (dewyatt) [inline src/librepgp/packet-print.c]: I think it makes sense to print the pubkey here too. I added `pgp_print_pubkey(printstate->indent, &seckey->pubkey);` locally and it produces some decent output (maybe a bit messy).

File: src/apps/packet-dumper/redumper.c [hunk 3]
```diff
@@ -0,0 +1,80 @@
+#include <stdio.h>
+#include <unistd.h>      /* getopt() */
+#include <repgp/repgp.h> /* repgp API */
+#include <rnp/rnp.h>     /* rnp_t, rnp_ctx_t et. all */
+#include <libgen.h>      /* basename() */
+
+#define PFX "redumper: "
+
+// extern int optind;
```
Comment (dewyatt) [inline src/apps/packet-dumper/redumper.c]: Can be removed I assume.

File: src/librepgp/packet-print.c (hunk 1 above)
Comment (flowher) [reply src/librepgp/packet-print.c]: Yup, got it. thnx

File: src/librepgp/packet-print.c (hunk 2 above)
Comment (flowher) [reply src/librepgp/packet-print.c]: Good point.

**Pull request merged**